src/private/MDMPerformerGroup.m

```diff
@@ -93,28 +100,55 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
 
 #pragma mark - Private
 
-- (id<MDMPerforming>)performerForPlan:(id<MDMPlan>)plan isNew:(BOOL *)isNew {
+- (void)removePlanNamed:(nonnull NSString *)name from:(nonnull id)target withPerformerInfo:(nullable MDMPerformerInfo *)performerInfo {
+  if (performerInfo) {
+    id<MDMPerforming> performer = performerInfo.performer;
+    if ([performer respondsToSelector:@selector(removePlanNamed:)]) {
+      [(id<MDMNamedPlanPerforming>)performer removePlanNamed:name];
+    }
+  } else {
+    // this is the case whereby the client has tried to remove a named performer which was never added in the first place
+  }
+}
+
+- (id<MDMPerforming>)findOrCreatePerformerForPlan:(id<MDMPlan>)plan trace:(MDMTrace *)trace isNew:(BOOL *)isNew {
+  return [self findOrCreatePerformerInfoForPlan:plan trace:trace isNew:isNew].performer;
+}
+
+- (MDMPerformerInfo *)findOrCreatePerformerInfoForNamedPlan:(id<MDMNamedPlan>)plan named:(NSString *)name trace:(MDMTrace *)trace isNew:(BOOL *)isNew {
+  // maybe this is a simple lookup
+  MDMPerformerInfo *performerInfo = self.performerPlanNameToPerformerInfo[name];
+  if (performerInfo) {
+    *isNew = NO;
+    return performerInfo;
+  }
+  // see if we can look it up by class name instead
+  performerInfo = [self findOrCreatePerformerInfoForPlan:plan trace:trace isNew:isNew];
+  // stash this perfomer info in case we ever want to look it up again
+  self.performerPlanNameToPerformerInfo[name] = performerInfo;
+  return performerInfo;
+}
+
+- (MDMPerformerInfo *)findOrCreatePerformerInfoForPlan:(id<MDMPlan>)plan trace:(MDMTrace *)trace isNew:(BOOL *)isNew {
   Class performerClass = [plan performerClass];
   id performerClassName = NSStringFromClass(performerClass);
   MDMPerformerInfo *performerInfo = self.performerClassNameToPerformerInfo[performerClassName];
   if (performerInfo) {
     *isNew = NO;
-    return performerInfo.performer;
+    return performerInfo;
+  } else {
+    id<MDMPerforming> performer = [[performerClass alloc] initWithTarget:self.target];
```

Comment:
- [ ] Remove the `} else {` clause here so that we have one less layer of scope. Don't remove the contents of the else clause. We lean on the fact that the if statement `returns`.